File: Data Structures & Algorithms/level-order-traversal-of-binary-tree/submission-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
from collections import deque
class Solution:
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        
        if not root:
            return []

        solution = []

        queue = deque()
        queue.append(root);

        level=0
        while len(queue) > 0:
            solution.append([])
            for i in range(len(queue)):
                node = queue.popleft()
                solution[level].append(node.val)
                if node.left:
                    queue.append(node.left)
                if node.right:
                    queue.append(node.right)
            level+=1
        return solution

File: Data Structures & Algorithms/balanced-binary-tree/submission-1.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    bool isBalanced(TreeNode* root) {
        bool res = true;
        dfs(root,res);
        return res;
    }

    int dfs(TreeNode* root, bool& res){
        if(!root){
            return 0;
        }

        int leftHeight = dfs(root->left,res);
        int rightHeight = dfs(root->right,res);
        if(abs(leftHeight-rightHeight)>1){
            res=false;
        }
        // cout<<leftHeight<<" - "<<rightHeight<<" - "<<res<<endl;
        return 1+max(leftHeight,rightHeight);
    }
};
